### ISSUE TITLE
Creating a partially connected layer

### ISSUE THREAD

**Issue by sunnsurf** [Microsoft/CNTK#3008]
I thought I could use convolution for this however I've so far been unable to identify the appropriate combination of shapes and parameters to achieve it so would greatly appreciate some guidance / insight please.

Layer input: 20 x 50 which represents a 50 sequences of values from different feature calculations
Target output: 5 x 50 (or 50 x 5) representing 5 weighted averages for each of the 50 feature calculations

My thought was to use 5 [20 x 1] kernels per feature calculation with a 20 x 1 stride however I'm yet to find the combination that executes (doesn't error) let alone produces the desired outcome. What is the appropriate way to achieve this outcome, and if using convolution isn't appropriate, what combination of C# functions is?

Simpler example: 3 x 4 => 2 x 4 (could also be 4 x 2)
```
[ a1 a2 a3           [  f1(a1,a2,a3), f2(a1,a2,a3)
  b1 b2 b3              f1(b1,b2,b3), f2(b1,b2,b3)
  c1 c2 c3       =>     f1(c1,c2,c3), f2(c1,c2,c3)
  d1 d2 d3 ]            f1(d1,d2,d3), f2(d1,d2,d3) ]
```

**Comment by KeDengMS** [comment]
In your simpler example, are the first two output values identical? If not, I think this is [convolution](https://cntk.ai/pythondocs/layerref.html#convolution) with num_filters = 5, and for your simpler example:
```
import cntk as C
x = C.sequence.input_variable((3)) # 4 is in the sequence axis, could be dynamic for each minibatch
y = C.squeeze(C.layers.Convolution1D(x.shape, num_filters=2, reduction_rank=0)(x)) # have the same sequence axis as x, and computes 2 weighted averages on 3 features per sample in x
import numpy as np
y.eval(np.random.rand(1,4,3)) # eval with batch_size of 1 (seq), seq_len of 4 (sample), sample shape of (3,)
```

**Comment by sunnsurf** [comment]
Both are functions of the same values although with different weights. I'll try to make it clearer.

**Comment by KeDengMS** [comment]
Yes then you should use num_filters as in the example code above.

**Comment by sunnsurf** [comment]
Input|Kernel|Stride|Result (expected total size: 8)
-----|------|------|------
[3, 1, 4] | [3, 1, 4, 2] | [3, 1, 4] | [1, 1, 2]
[3, 4] | [3, 1, 1, 2] | [3, 4] | [1, 2]
[3, 4] | [3, 1, 2] | [3, 4] |  [1, 2]
[3, 4] | [3, 1, 2] | [3, 1] | Validating --> Convolution3 = Convolution (Parameter2, Input0) : [3 x 1 x 2], [3 x 4 x *2] -> [1 x 4 x *2] FAILED

**Comment by KeDengMS** [comment]
OK so 4 is part of sample shape which you don't want convolution to run on. Your kernelDepth should be 1 in this case. Please try input shape of [3,4], Kernel shape [3, 1, 1, 2], stride [1,1] with no padding. The output shape [1,2] because the reduced dimension of 3 becomes 1, and you can use reshape or squeeze to make the shape [2].

For detailed explanation of how convolution works, please take a look at [tutorial 103D](https://github.com/Microsoft/CNTK/blob/master/Tutorials/CNTK_103D_MNIST_ConvolutionalNeuralNetwork.ipynb).

**Comment by sunnsurf** [comment]
Input [3,4] with Kernel [3,1,1,2] and Stride [1,1] still results in [1,2] instead of total size of 8 (4*2).

**Comment by KeDengMS** [comment]
I tried following code in latest CNTK:
```
        private static readonly Variable _featureVariable =
            Variable.InputVariable(new[] { 3, 4 }, DataType.Float, "features");

        private static readonly DeviceDescriptor device = DeviceDescriptor.CPUDevice; // <--- CPU

        public static void Main()
        {
            var convParams = new Parameter(new int[] { 3, 1, 2 },
                                            DataType.Float,
                                            CNTKLib.GlorotUniformInitializer(0.01, -1, 2, 0),
                                            device);

            Function conv = CNTKLib.Squeeze(CNTKLib.Convolution(convParams,
                                               _featureVariable,
                                               /* strides */ new int[] {1, 1},
                                               /* shared */ new CNTK.BoolVector { false },
                                               /* padding */ new CNTK.BoolVector { false },
                                               /* dilation */ new int[] { 1, 1 },
                                               /* reduction rank */ 0));

            System.Console.WriteLine(conv.AsString());
        }
```
And the output is
```
Composite(Squeeze): Input('features', [3 x 4], [*, #]) -> Output('Squeeze19_Output_0', [4 x 2], [*, #])
```

**Comment by sunnsurf** [comment]
The error occurs on the call to Trainer

```C#
private static Random r = new Random();

public static Function FullyConnectedLinearLayer(Variable input, int outputDim, DeviceDescriptor device, string outputName = "")
{
    if (input.Shape.Rank != 1)
    {
        input = CNTKLib.Reshape(input, new int[] { input.Shape.TotalSize });
    }
    
    var timesParam = new Parameter(new int[] { outputDim, input.Shape[0] },
            DataType.Float,
            CNTKLib.GlorotUniformInitializer(
                CNTKLib.DefaultParamInitScale,
                CNTKLib.SentinelValueForInferParamInitRank,
                CNTKLib.SentinelValueForInferParamInitRank,
                (uint)r.Next()),
            device);
    
    var plusParam = new Parameter(new int[] { outputDim },
                            DataType.Float,
                            CNTKLib.GlorotUniformInitializer(
                                CNTKLib.DefaultParamInitScale,
                                CNTKLib.SentinelValueForInferParamInitRank,
                                CNTKLib.SentinelValueForInferParamInitRank,
                                (uint)r.Next()),
                            device);

    return CNTKLib.Plus(plusParam, CNTKLib.Times(timesParam, input), outputName);
}

//---
static void Main(string[] args)
{

    Variable inputVariable = Variable.InputVariable(new[] { 3, 4 }, DataType.Float, "in");
    float[] inputData = new float[24];
    for (uint i = 0; inputData.Length > i; i++)
        inputData[i] = r.Next(100) / 100f;
    Value inputValue = Value.CreateBatch(inputVariable.Shape, inputData, 0, inputData.Length, DeviceDescriptor.CPUDevice, /* readonly = */ false);


    Variable outputVariable = Variable.InputVariable(new[] { 2 }, DataType.Float, "out");
    float[] outputData = new float[4] { 0, 1, 1, 0 };
    Value outputValue = Value.CreateBatch(outputVariable.Shape, outputData, 0, outputData.Length, DeviceDescriptor.CPUDevice, /* readonly = */ false);

    var convParams = new Parameter(new int[] { 3, 1, 2 },
                                    DataType.Float,
                                    CNTKLib.GlorotUniformInitializer(0.01, -1, 2, 0),
                                    DeviceDescriptor.CPUDevice);

    Function conv = CNTKLib.Squeeze(CNTKLib.Convolution(convParams,
                                       inputVariable,
                                       /* strides */ new int[] { 1, 1 },
                                       /* shared */ new CNTK.BoolVector { false },
                                       /* padding */ new CNTK.BoolVector { false },
                                       /* dilation */ new int[] { 1, 1 },
                                       /* reduction rank */ 0));

    System.Console.WriteLine(conv.AsString());

    Function classifier = FullyConnectedLinearLayer(conv, 2, DeviceDescriptor.CPUDevice, "Classifier");

    ParameterVector parameterVector = new ParameterVector(classifier.Parameters().ToList());
    TrainingParameterScheduleDouble learningRateSchedule = new TrainingParameterScheduleDouble(0.002, /* batch size */ 2);        
    TrainingParameterScheduleDouble momentumRateSchedule = new TrainingParameterScheduleDouble(0.9, /* batch size */ 2);          
    TrainingParameterScheduleDouble varianceMomentumSchedule = new TrainingParameterScheduleDouble(0.999, /* batch size */ 2);    
    //---
    IList<Learner> parameterLearners = new List<Learner>()
    {
            CNTKLib.AdamLearner(parameterVector, learningRateSchedule, momentumRateSchedule, /* unitGainMomentum = */true)
    };

    Function TrainingLoss = CNTKLib.CrossEntropyWithSoftmax(classifier, outputVariable, "lossFunction");
    Function ClassificationError = CNTKLib.ClassificationError(classifier, outputVariable, "classificationError");
    
    /*
     * Error is here
     */
    Trainer trainer = Trainer.CreateTrainer(classifier, TrainingLoss, ClassificationError, parameterLearners);
    
    Dictionary<Variable, Value> Data = new Dictionary<Variable, Value>
        {
            { inputVariable, inputValue },
            { outputVariable, outputValue }
        };
    trainer.TrainMinibatch(Data, /* sweep end */ true, DeviceDescriptor.CPUDevice);

    float xe = (float)trainer.PreviousMinibatchLossAverage();
    float ce = (float)trainer.PreviousMinibatchEvaluationAverage();
    Console.WriteLine($"xe = {xe}; ce = {ce}");
}
```

Results in 
```
Composite(Squeeze): Input('in', [3 x 4], [*, #]) -> Output('Squeeze20_Output_0', [4 x 2], [*, #])
Validating --> Convolution10 = Convolution (Reshape4, Reshape7) : [3 x 1 x 1 x 2], [3 x 4 x 1 x *2] -> [1 x 4 x 2 x *2] FAILED

System.ApplicationException
  HResult=0x80131600
  Message=Convolution weight matrix Reshape4 should have dimension [(filter shape) x (input channels) x (output channels)]

[CALL STACK]
    > std::enable_shared_from_this<Microsoft::MSR::CNTK::MatrixBase>::  shared_from_this
    - std::enable_shared_from_this<Microsoft::MSR::CNTK::MatrixBase>::  shared_from_this (x5)
    - CNTK::Internal::  UseSparseGradientAggregationInDataParallelSGD (x3)
    - CNTK::Function::  Forward
    - CNTK::  CreateTrainer
    - CNTK::Trainer::  TotalNumberOfUnitsSeen
    - CNTK::Trainer::  TrainMinibatch (x2)
    - CSharp_CNTK_Trainer__TrainMinibatch__SWIG_2
    - 00007FFEA7521525 (SymFromAddr() error: The specified module could not be found.)


  Source=Cntk.Core.Managed-2.4
  StackTrace:
   at CNTK.Trainer._TrainMinibatch(UnorderedMapVariableValuePtr arguments, Boolean isSweepEndInArguments, DeviceDescriptor computeDevice)
```

**Comment by KeDengMS** [comment]
My bad, the sharing should be true instead of false. Try this instead:
```
            Function conv = CNTKLib.Squeeze(CNTKLib.Convolution(convParams,
                                               _featureVariable,
                                               /* strides */ new int[] {1, 1},
                                               /* shared */ new CNTK.BoolVector { true },
                                               /* padding */ new CNTK.BoolVector { false },
                                               /* dilation */ new int[] { 1, 1 },
                                               /* reduction rank */ 0));
```

**Comment by sunnsurf** [comment]
Thanks @KeDengMS, you're a champion. It appears to be running now. I can increase the number of filters for now as workaround for sharing kernels. Just to confirm my understanding, if the number of filters is 2 as per our simple example, does that mean there are 2 weight kernels? That is, although the 4 variables are treated separately, they still share the same 3 weights and bias values?

**Comment by KeDengMS** [comment]
Yes num_filters = 2 means your have 2 kernels, each with dim of 3. The 4 locations for convolution shares the weights and bias of those filters.
